kernel/src/scan/mod.rs

```diff
@@ -77,18 +78,29 @@ impl ScanBuilder {
 
     /// Build the [`Scan`].
     ///
-    /// This is lazy and performs no 'work' at this point. The [`Scan`] type itself can be used
-    /// to fetch the files and associated metadata required to perform actual data reads.
-    pub fn build(self) -> Scan {
+    /// This does not scan the table at this point, but does do some work to ensure that the
+    /// provided schema make sense, and to prepare some metadata that the scan will need.  The
+    /// [`Scan`] type itself can be used to fetch the files and associated metadata required to
+    /// perform actual data reads.
+    pub fn build(self) -> DeltaResult<Scan> {
         // if no schema is provided, use snapshot's entire schema (e.g. SELECT *)
-        let read_schema = self
+        let logical_schema = self
             .schema
             .unwrap_or_else(|| self.snapshot.schema().clone().into());
-        Scan {
+        let (all_fields, read_fields, have_partition_cols) = get_state_info(
```

Comment:
aside: This takes a lot of `self.snapshot` args... should it be `Snapshot::get_scan_state_info` or similar?

Reply:
maybe? is 2 a lot? :) 

`Snapshot::get_scan_state_info` would need to take a logical schema, so I'm not clear it's a win

Reply:
Yeah, you're probably right. I was just thinking out loud